File: android/sample-ui/src/main/kotlin/com/usesmileid/sampleapps/ui/components/UseSmileIDSampleSelectionCheckbox.kt
package com.usesmileid.sampleapps.ui.components

import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.selection.toggleable
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Surface
import androidx.compose.material3.minimumInteractiveComponentSize
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.unit.dp
import com.smileid.designsystem.SmileDimens
import com.smileid.designsystem.smileBorderStrong
import com.usesmileid.sampleapps.ui.theme.UseSmileIDSampleTheme

/** The select-mode checkbox. Circular rather than a rounded square, which is the design's correction. */
@Composable
fun UseSmileIDSampleSelectionCheckbox(
    checked: Boolean,
    onCheckedChange: (Boolean) -> Unit,
    modifier: Modifier = Modifier,
    testId: String? = null,
) {
    val colors = UseSmileIDSampleTheme.colors
    Box(
        modifier = modifier
            .minimumInteractiveComponentSize()
            .toggleable(value = checked, role = Role.Checkbox, onValueChange = onCheckedChange)
            .tagged(testId),
        contentAlignment = Alignment.Center,
    ) {
        Surface(
            modifier = Modifier
                .size(SmileDimens.sizeIconLg)
                .border(
                    // A 2px ring in border-strong: color.border is far too pale to read as a control.
                    width = SmileDimens.borderWidthThick,
                    color = if (checked) colors.primary else smileBorderStrong,
                    shape = CircleShape,
                ),
            shape = CircleShape,
            color = if (checked) colors.primary else colors.surface,
        ) {
            Box(contentAlignment = Alignment.Center) {
                if (checked) CheckGlyph(tint = colors.onPrimary, size = CHECK_SIZE)
            }
        }
    }
}

private val CHECK_SIZE = 11.dp

File: flutter/sample_ui/lib/src/components/use_smileid_sample_selection_checkbox.dart
import 'package:flutter/material.dart';

import '../theme/use_smileid_sample_colors.dart';
import '../theme/use_smileid_sample_theme.dart';
import '../tokens/smile_product_hues.dart';
import '../tokens/smile_tokens.dart';
import 'use_smileid_sample_glyphs.dart';

/// The select-mode checkbox: circular rather than a rounded square, which is the design's correction.
class UseSmileIDSampleSelectionCheckbox extends StatelessWidget {
  /// Takes the value rather than owning it, so the screen's selection set is the single truth.
  const UseSmileIDSampleSelectionCheckbox({
    required this.checked,
    required this.onChanged,
    this.testId,
    super.key,
  });

  /// Whether the row is selected.
  final bool checked;

  /// Called with the new value.
  final ValueChanged<bool> onChanged;

  /// The `sample_*` id the screen supplies.
  final String? testId;

  @override
  Widget build(BuildContext context) {
    final UseSmileIDSampleColors colors = UseSmileIDSampleTheme.colorsOf(
      context,
    );
    return Semantics(
      identifier: testId,
      checked: checked,
      child: InkResponse(
        onTap: () => onChanged(!checked),
        radius: SmileDimens.sizeControlMd / 2,
        child: SizedBox(
          // The platform interactive minimum around a 24 control, not a 24 tap target.
          width: SmileDimens.sizeControlMd,
          height: SmileDimens.sizeControlMd,
          child: Center(
            child: Container(
              width: SmileDimens.sizeIconLg,
              height: SmileDimens.sizeIconLg,
              alignment: Alignment.center,
              decoration: BoxDecoration(
                shape: BoxShape.circle,
                color: checked ? colors.primary : colors.surface,
                // A 2 ring in border-strong: color.border is far too pale to read as a control.
                border: Border.all(
                  width: SmileDimens.borderWidthThick,
                  color: checked ? colors.primary : smileBorderStrong,
                ),
              ),
              child: checked
                  ? UseSmileIDSampleGlyphs.check(
                      colors.onPrimary,
                      size: _checkSize,
                    )
                  : null,
            ),
          ),
        ),
      ),
    );
  }
}

/// The glyph inside a 24 ring; the icon scale's 16 leaves no ring visible.
const double _checkSize = 11;
